pkg/geo/geomfn/distance.go

```diff
@@ -865,10 +868,14 @@ func PointKindRelatesToPolygonKind(
 	pointKindIterator := geo.NewGeomTIterator(pointKindBaseT, geo.EmptyBehaviorOmit)
 	polygonKindIterator := geo.NewGeomTIterator(polygonKindBaseT, geo.EmptyBehaviorOmit)
 
+	// TODO(ayang): Think about how to refactor these nested for loops
```

Comment:
I think these for loops have gotten so unwieldy to the point it may be better to factor out the common setup code and have a separate nested for loop per predicate. Also, maybe all the point in polygon optimization code should be moved to a new file called `point_in_polygon.go` or something. @otan What do you think?

Reply:
Happy to move this into a separate file.

Also I think you can do something similar to DistanceUpdater/Distance to make it generic.. I think the interface you need to do something similar is:

```
type PointInPolygonEventListener struct {
  // this function is injected because the rest of the logic is dependent on of "inside", "outside" or "on" linear ring
  IntersectsFunc(*geom.Polygon, *geom.Point) bool 

   // bool returns whether to early exit and return true 
   OnShapeIntersects() bool
   // bool returns whether to early exit and return false
   OnShapeDoesNotIntersect() bool
}
```

Reply:
Thanks for the suggestion, I'll give that a try in a separate PR